src/tools/cephfs/DataScan.cc

```diff
@@ -0,0 +1,1267 @@
+        uint64_t max_obj_id;
+        uint64_t max_obj_size;
+        ::decode(max_obj_id, scan_size_bl_iter);
+        ::decode(max_obj_size, scan_size_bl_iter);
+
+        if (max_obj_id > 0) {
+          // If there are more objects, and this object is a power of two
+          // that is greater than the size of the last object, then make
+          // an educated guess that this object's size is probably the chunk
+          // size.
+        
+          // TODO: provide a means for the user to specify the layout, as
+          // for e.g. striped layouts there is no way we can infer it.
+          if ((obj_size & (obj_size - 1)) == 0 && obj_size >= max_obj_size) {
+            chunk_size = obj_size;
+          } else {
+            chunk_size = g_default_file_layout.fl_object_size;
+          }
+          file_size = chunk_size * max_obj_id + max_obj_size;
+        } else {
+          file_size = obj_size;
+        }
+      } catch (const buffer::error &err) {
+        dout(4) << "Invalid size attr on '" << oid << "'" << dendl;
+        file_size = obj_size;
+      }
+    } else {
+      file_size = obj_size;
+    }
+
+    // Read accumulated mtime from scan_extents phase
+    time_t file_mtime = 0;
+    bufferlist scan_mtime_bl;
+    r = data_io.getxattr(oid, "scan_mtime", scan_mtime_bl);
+    if (r >= 0) {
+      try {
+        bufferlist::iterator scan_mtime_bl_iter = scan_mtime_bl.begin();
+        ::decode(file_mtime, scan_mtime_bl_iter);
+      } catch (const buffer::error &err) {
+        dout(4) << "Invalid mtime attr on '" << oid << "'" << dendl;
+        file_mtime = obj_mtime;
+      }
+    } else {
+      file_mtime = obj_mtime;
+    }
+
+    // Inject inode to the metadata pool
+    if (have_backtrace) {
+      // TODO: inspect backtrace and fall back to lost+found
+      // for some or all stray cases.
+      r = driver->inject_with_backtrace(
+          backtrace, file_size, file_mtime, chunk_size, data_pool_id);
+      if (r < 0) {
+        dout(4) << "Error injecting 0x" << std::hex << backtrace.ino
+          << std::dec << " with backtrace: " << cpp_strerror(r) << dendl;
+      }
+    } else {
+      r = driver->inject_lost_and_found(
+          obj_name_ino, file_size, file_mtime, chunk_size, data_pool_id);
+      if (r < 0) {
+        dout(4) << "Error injecting 0x" << std::hex << obj_name_ino
+          << std::dec << " into lost+found: " << cpp_strerror(r) << dendl;
+      }
+    }
+  }
+
+  return 0;
+}
+
+int MetadataDriver::read_fnode(inodeno_t ino, frag_t frag, fnode_t *fnode)
+{
+  assert(fnode != NULL);
+
+  object_t frag_oid = InodeStore::get_object_name(ino, frag, "");
+  bufferlist old_fnode_bl;
+  int r = metadata_io.omap_get_header(frag_oid.name, &old_fnode_bl);
+  if (r < 0) {
+    return r;
+  }
+
+  bufferlist::iterator old_fnode_iter = old_fnode_bl.begin();
+  try {
+    fnode_t old_fnode;
+    old_fnode.decode(old_fnode_iter);
+  } catch (const buffer::error &err) {
+    return -EINVAL;
+  }
```

Comment:
Shouldn't we be copying the fnode into the output pointer one here? :/
(Or just using it in place of this old_fnode local member.)